### PR TITLE
[ray] Update cpp to std14

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Updating cpp from std-11 to std-14. make_unique added to make sure it's working.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:1490]: only to make sure it's working

Comment (qicosmos) [reply src/ray/core_worker/core_worker.cc:1490]: If we support c++14 we can get more benefits, we can optimize performance more easier.
We can use lambda capture move in c++14:
```c++
  std::string str = "test";
  //c++11, we have to copy here.
  auto lambda11 = [str]{
  };

  //c++14, move the str, no copy anymore
  auto lambda14 = [value = std::move(str)]{
  };
```

You can also use generic lambda  and auto function to simplify your code:
```c++
//generic lambda, it's very useful to simplify template code.
auto lambda = [](auto x, auto y) {return x + y;};

///auto function to cut the tail in c++11

//in c++11
template<typename F, typename Tuple, int... I>
auto apply_(F&& f, Tuple&& args, int_seq<I...>) ->
  decltype(std::forward<F>(f)(std::get<I>(std::forward<Tuple>(args))...)) //auto-decltype tail in c++11
  {
    return std::forward<F>(f)(std::get<I>(std::forward<Tuple>(args))...);
  }

template <typename F, typename Tuple, size_t... I>
decltype(auto) apply_impl(F&& f, Tuple&& t, index_sequence<I...>) { //no tail anymore
    return forward<F>(f)(get<I>(forward<Tuple>(t))...);
}
```

There also some other cool features in c++14, such as template variable, more useful type traits, actually the ci has run on gcc 5.4 which has already fully supported c++14, i really hope we can update to c++14.

Today is 2021, c++20 published last year, but we are using the c++ standard 10 years ago, we should update at least c++14.

c++17 and c++20 are also very good to own:)

**Comment by rkooo567** [comment]
According to the test result, it seems like the transition would be smooth!

**Review by ericl — COMMENTED** [review]
LGTM but let's make sure the wheels all build. You can label test-ok once that's confirmed

**Comment by pcmoritz** [comment]
LGTM once the wheels pass (and we should also send them out for manual testing on different macOS and linux versions to make sure they work)

**Comment by qicosmos** [comment]
> LGTM once the wheels pass (and we should also send them out for manual testing on different macOS and linux versions to make sure they work)

If there are some compile errors, please paste the errors here, i'm glad to do some to fix compile errors.

**Comment by iycheng** [comment]
@pcmoritz  It looks like one of the builds keeps timeout even after I increase it to 500 and merge to trunk.

With the help from @simon-mo we feel it's might due to the cache missing. Considering this flag will make all caching invalid, it makes sense. One thing we can try is to push it to see whether that test case still fails to build.

**Comment by iycheng** [comment]
Synced with @pcmoritz and we can check it in first to see how things going. If it's not working well, we can revert it. Could you help merge it @ericl

**Review by ericl — APPROVED** [review]

**Comment by iycheng** [comment]
I checked the recent builds and I feel we are good here!!

**Comment by kfstorm** [comment]
I'm not sure if they are related.

image

**Pull request merged**
